vignettes/articles/oauth.Rmd

```diff
@@ -14,24 +14,61 @@ library(httr2)
+This means that while httr2 can make the interactive experience pretty seamless, unattended usage (i.e. on CI or in an automated script) is usually going to include some pain.
+
+This vignette builds on the techniques in discussed in `vignette("wrapping-apis")`, so I recommend you read that vignette first.
+I'll also assume that you're working in package, since most people want to wrap up an API into a bunch of R functions, but most of the ideas will also apply if you're creating a one off script.
+
+## OAuth basics
+
+OAuth is a broad framework that has many different variants, called **flows**, which makes it hard to provide sweeping generalisations, but the basic idea of OAuth is to create a hierarchy of increasing more specific and shorter lived credentials, so that the impact of a credential being lost is as small as possible.
+
+The longest lived and most powerful credential is typically a user name and password.
+Most people don't change their passwords regularly, and often (against all advice) reuse the same password on multiple websites.
+And if you have a user name and password, you have total control over that account; you can even use it to change the password so the actual user can't log in.
+That means as a programmer you never want to touch user name-password pairs because if they're lost or stolen, they give very wide access.
+
+Avoiding that problem lead to the creation of OAuth.
+The basic idea is that instead of your package asking the user to give you their user name and password, you instead ask them to log in and give your package permission to use the API on their behalf.
+The API gives this permission in the form of an **access token**, which is essentially a random string of numbers and letters.
+An access token tends to be very short lived, lasting maybe only days or hours, and is bound to a specified **scope** of access.
+The access token has some big advantages over a user name and password:
+
+-   It's short-lived so if it's lost or stolen, there's only a limited amount of time where it can be abused.
+-   It has limited scope, so even if stolen, it can't be used to do something particularly nefarious like changing your password or contact details.
+-   It's bound to a specific application, so it can be invalidated (cancelled) without affecting any other uses.
+
+If you have an access token you can use it to authenticate an API by passing a **bearer token** in the `Authorization` header, which you can do with `req_auth_bearer_token()`.
+However, in most cases you will want to let httr2 manage this by calling one of the `req_oauth_` functions we'll talk about shortly.
+
+One of the reasons that you want httr2 to manage your tokens is that because access tokens are so short lived, they're often accompanied by a **refresh token**.
+A refresh token lasts for a longer amount of time and has one job: it allows you to get a new access token when the previous one expires.
```

Comment:
to me that's always been confusing because losing the refresh token seems dangerous so I'd appreciate some comment on that.

Reply:
Don’t lose it 😀